### PR TITLE
iroh: 0.11 -> 0.12

### PR DESCRIPTION
**Pull request by crimeminister**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by crimeminister** [comment]
An error compiling a dependency is currently blocking:
```_
error: couldn't read /build/iroh-0.12.0-vendor.tar.gz/watchable/src/.crate-docs.md: No such file or directory (os error 2)
 --> /build/iroh-0.12.0-vendor.tar.gz/watchable/src/lib.rs:1:10
  |
1 | #![doc = include_str!(".crate-docs.md")]
  |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by cameronfyfe** [comment]
> An error compiling a dependency is currently blocking:
> 
> ```
> error: couldn't read /build/iroh-0.12.0-vendor.tar.gz/watchable/src/.crate-docs.md: No such file or directory (os error 2)
>  --> /build/iroh-0.12.0-vendor.tar.gz/watchable/src/lib.rs:1:10
>   |
> 1 | #![doc = include_str!(".crate-docs.md")]
>   |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> ```

It looks like `.crate-docs.md` in this `watchable` crate used by v0.12.0 for some reason gets filtered out by the `rustPlatform.fetchCargoTarball` logic which calls cargo vendor` (I ran `cargo vendor` manually for `watchable` and that `.crate-docs.md` is missing in the resulting vendored folder). `.crate-docs.md` **is** in the published package for `watchable` though and since `rustPlatform.importCargoLock` doesn't use `cargo vendor` the build seems to work if you switch to that. It's also nice to switch to this so a hash for the dependencies doesn't need to be manually specified anymore. Putting working code in separate comment.

File: pkgs/applications/networking/iroh/default.nix
Comment (cameronfyfe) [inline pkgs/applications/networking/iroh/default.nix:19]:
```suggestion
  cargoLock = {
    lockFile = "${src}/Cargo.lock";
  };
```

**Comment by cameronfyfe** [comment]
@crimeminister I forgot that you're not allowed to import from derivation so the `lockFile = "${src}/Cargo.lock"` I recommended is not allowed. You can remedy this by copying and commiting the whole `Cargo.lock` file here (this is what other people are doing) or you can wait until the root issue is fixed in the next version of iroh and go back to specifying `cargoHash` (I submitted this PR https://github.com/n0-computer/iroh/pull/1932).

**Comment by crimeminister** [comment]
> not allowed. You can remedy this by copying an

Thanks so much for your investigation @cameronfyfe. I'll include the `Cargo.lock` file for now, and track the progress of your PR with the plan of sending another update for the next version, assuming you don't get there first.

**Review by cameronfyfe — APPROVED** [review]
Works on my machine (x86-linux)

**Comment by crimeminister** [comment]
@cameronfyfe Is there another step to getting this PR reviewed and/or merged? Thanks.

**Comment by cameronfyfe** [comment]
@crimeminister Just need to wait for someone with commit rights to approve/merge it, I can only approve it. Sometimes takes a bit. If you want it soon you can try to find someone to request review from or post about it in a forum and hope someone sees it.

**Comment by crimeminister** [comment]
> @crimeminister Just need to wait for someone with commit rights to approve/merge it, I can only approve it. Sometimes takes a bit. If you want it soon you can try to find someone to request review from or post about it in a forum and hope someone sees it.

Thanks @cameronfyfe. I am in no particular rush, I just wanted to make sure I wasn't missing a step in the process on my end.

**Comment by wegank** [comment]
Done in #299006.

**Pull request closed without merging**
